File: 23_设计类/355. 设计推特.ts
import { Deque } from '../2_queue/Deque/ArrayDeque'
import { PriorityQueue } from '../2_queue/优先级队列'

function* count() {
  let count = 0
  while (true) {
    yield count++
  }
}

class Timer {
  private timer: Generator<number>

  constructor() {
    this.timer = count()
  }

  next(): number {
    return this.timer.next().value
  }
}

type FollowerId = number
type FolloweeId = number
type UserId = number
interface Tweet {
  id: number
  time: number
}

class Twitter {
  private followMap: Map<FollowerId, Set<FolloweeId>>
  private tweetMap: Map<UserId, Deque<Tweet>> // 这里其实最好使用deque
  private timer: Timer

  constructor() {
    this.followMap = new Map()
    this.tweetMap = new Map()
    this.timer = new Timer()
  }

  /**
   *
   * @param userId
   * @param tweetId
   * 创建一条新的推文
   */
  postTweet(userId: number, tweetId: number): void {
    !this.tweetMap.has(userId) && this.tweetMap.set(userId, new Deque(10))
    this.tweetMap.get(userId)!.unshift({ id: tweetId, time: this.timer.next() })
  }

  /**
   *
   * @param userId
   * 检索最近的十条推文。每个推文都必须是由此用户关注的人或者是用户自己发出的。
   * 推文必须按照时间顺序由最近的开始排序。
   * @summary
   * 使用pq合并 k 个有序链表
   */
  getNewsFeed(userId: number): number[] {
    const allTweets = [...(this.followMap.get(userId) || []), userId]
      .filter(id => this.tweetMap.has(id))
      .map(id => this.tweetMap.get(id)!)

    const priorityQueue = new PriorityQueue<Tweet>((a, b) => b.time - a.time)

    allTweets.forEach(tweets => tweets.forEach(tweet => priorityQueue.push(tweet)))

    priorityQueue.heapify()

    const res: number[] = []
    while (priorityQueue.length && res.length < 10) {
      res.push(priorityQueue.shift()!.id)
    }

    return res
  }

  /**
   *
   * @param followerId
   * @param followeeId
   * 关注一个用户
   */
  follow(followerId: number, followeeId: number): void {
    !this.followMap.has(followerId) && this.followMap.set(followerId, new Set())
    this.followMap.get(followerId)!.add(followeeId)
  }

  /**
   *
   * @param followerId
   * @param followeeId
   * 取消关注一个用户
   */
  unfollow(followerId: number, followeeId: number): void {
    if (!this.followMap.has(followerId)) return
    this.followMap.get(followerId)!.delete(followeeId)
  }
}

const twitter = new Twitter()

// 用户1发送了一条新推文 (用户id = 1, 推文id = 5).
twitter.postTweet(1, 5)
twitter.postTweet(1, 3)
twitter.postTweet(1, 101)
twitter.postTweet(1, 13)
twitter.postTweet(1, 10)
twitter.postTweet(1, 2)
twitter.postTweet(1, 94)
twitter.postTweet(1, 505)
twitter.postTweet(1, 333)

// 用户1的获取推文应当返回一个列表，其中包含一个id为5的推文.
console.log(twitter.getNewsFeed(1))

// 用户1关注了用户2.
// console.log(twitter.follow(1, 2))

// // 用户2发送了一个新推文 (推文id = 6).
// console.log(twitter.postTweet(2, 6))

// // // 用户1的获取推文应当返回一个列表，其中包含两个推文，id分别为 -> [6, 5].
// // // 推文id6应当在推文id5之前，因为它是在5之后发送的.
// console.log(twitter.getNewsFeed(1))
// // // 用户1取消关注了用户2.
// twitter.unfollow(1, 2)

// // // 用户1的获取推文应当返回一个列表，其中包含一个id为5的推文.
// // // 因为用户1已经不再关注用户2.
// console.log(twitter.getNewsFeed(1))

File: 23_设计类/355. 设计推特.py
"""
推特的两个主要业务是：

1.发布推文(postTweet) 写
  用户可以向其粉丝发布新消息（平均 4.6k 请求 / 秒，峰值超过 12k 请求 / 秒）。

2.主页时间线(getNewsFeed) 读
  用户可以查阅他们关注的人发布的推文(300k 请求 / 秒）。

处理每秒 12,000 次写入（发推文的速率峰值）还是很简单的。
推特的伸缩性挑战并不是主要来自推特量，而是来自扇出(fan-out,为了服务一个传入请求而需要执行其他服务的请求数量。)

这一对操作有两种实现方式。
1. 读时计算量大:
发布推文时直接将推文存入用户推文的哈希表，当一个用户请求自己的主页时间线时，
首先查找他关注的所有人，查询这些被关注用户发布的推文并按时间顺序合并

2. 写时计算量大:
为每个用户的主页时间线维护一个缓存，就像每个用户的推文收件箱
当一个用户发布推文时，查找所有关注该用户的人，并将新的推文插入到每个人的推文收件箱
因此读取主页时间线的请求开销很小，因为结果已经提前计算好了。

ps:设计成deque便于删除队头已取消关注者的文章/删除队尾过期的文章(如果需要这个功能的话)

推特的第一个版本使用了方法 1,但系统很难跟上主页时间线查询的负载。所以公司转向了方法 2
方法 2 的效果更好，因为发推频率比查询主页时间线的频率几乎低了两个数量级
然而方法 2 的缺点是 一些用户有超过 3000 万的粉丝，这意味着一条推文就可能会导致主页时间线缓存的 3000 万次写入
在推特的例子中，每个用户粉丝数的分布（可能按这些用户的发推频率来加权）是探讨可伸缩性的一个关键负载参数，因为它决定了扇出负载
推特轶事的最终转折：现在已经稳健地实现了方法 2,推特逐步转向了两种方法的混合
即:当用户读取主页时间线时，分别地获取出该用户所关注的每位名流的推文，再与用户的主页时间线缓存合并
"""

from typing import List
from collections import defaultdict, deque
from heapq import merge
from itertools import count, islice


class Twitter:
    def __init__(self):
        self.timer = count(step=-1)
        self.tweet = defaultdict(deque)
        self.following = defaultdict(set)

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.tweet[userId].appendleft((next(self.timer), userId, tweetId))

    def getNewsFeed(self, userId: int) -> List[int]:
        """时间复杂度O(klog(n)) 这里k为10"""
        tweets = merge(
            *(self.tweet[u] for u in self.following[userId] | {userId})
        )  # 注意| {userId} 需要并上自己的推文
        return [t for *_, t in islice(tweets, 10)]  # !至多取前10项

    def follow(self, followerId: int, followeeId: int) -> None:
        self.following[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        self.following[followerId].discard(followeeId)


class _Twitter2:
    def __init__(self):
        self.timer = count(step=-1)
        self.tweet = defaultdict(deque)
        self.feedLine = defaultdict(deque)  # 每个用户的推文收件箱
        self.following = defaultdict(set)  # 自己关注了谁
        self.follower = defaultdict(set)  # 谁关注了自己

    def postTweet(self, userId: int, tweetId: int) -> None:
        """当一个用户发布推文时，查找所有关注该用户的人，并将新的推文插入到每个人的推文收件箱"""
        newTweet = (next(self.timer), userId, tweetId)
        self.tweet[userId].appendleft(newTweet)
        for user in self.follower[userId] | {userId}:  # 注意需要并上自己
            self.feedLine[user].appendleft(newTweet)

    def getNewsFeed(self, userId: int) -> List[int]:
        """每个用户的主页时间线维护一个缓存，就像每个用户的推文收件箱;如果用户已经取关,那么就删除"""
        line, res = self.feedLine[userId], []
        while line and len(res) < 10:  # !至多取前10项
            head = line.popleft()
            postUser = head[1]
            if postUser in self.following[userId] | {userId}:  # 注意需要并上自己
                res.append(head)
        line.extendleft(res[::-1])
        return [t for *_, t in res]

    def follow(self, followerId: int, followeeId: int) -> None:
        """注意这种设计 follow时followeeId之前的动态不会出现在自己的line里"""
        self.following[followerId].add(followeeId)
        self.follower[followeeId].add(followerId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        """unfollow后在line里延迟删除"""
        self.following[followerId].discard(followeeId)
        self.follower[followeeId].discard(followerId)


if __name__ == "__main__":
    test = [(-1, 1), (0, 1)]
    print([(a, b) for a, b in islice(test, 3)])
    pq = [12, 3, 66]
    print(*merge(pq, [2, 3, 4]))
